feat(argo-workflows): added new argo-workflows chart and deprecated argo chart

chgl opened the pull request:
Resolves #660:

uses the workaround of setting `--secure=false` by default: https://github.com/argoproj/argo-workflows/issues/5607

resolves #646:

renames the chart (while keeping the old one around but marking as deprecated)

resolves #636:

copied all the CRDs from https://github.com/argoproj/argo-workflows/tree/master/manifests/base/crds/minimal and updated clusterroles for the new workfloweventbindings resource

resolves #483:

while the apiextension version was bumped to v1, the resource spec was not updated correctly. This is solved by resolving #636 as well.

I've documented the breaking changes in the README. I tried not to change *too* much, but we could definitely break more - now is the time!

I've removed any included minio support for now until https://github.com/argoproj/argo-helm/pull/649 is resolved - so we can add it back later with a clean solution.

Checklist:

* [x] I have updated the chart version in `Chart.yaml` following Semantic Versioning.
* [ ] Any new values are backwards compatible and/or have sensible default.
* [x] I have followed the testing instructions in the [contributing guide](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md).
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo).
* [x] My build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

Changes are automatically published when merged to `master`. They are not published on branches.

ChiefAlexander commented on charts/argo-workflows/templates/controller/workflow-controller-deployment.yaml:
```diff
@@ -0,0 +1,114 @@
+          {{- if .Values.controller.metricsConfig.enabled }}
+          ports:
+            - name: metrics
+              containerPort: {{- .Values.controller.metricsConfig.port }}
```
You need to remove the `-` to avoid templating issues.

```suggestion
              containerPort: {{ .Values.controller.metricsConfig.port }}
```

chgl replied:
argh, thanks for noticing. Will fix asap. In general, I am not too happy with the way the metricsConfig currently works - I kept the changes minimal but I think it may be better to enable the metrics port at all times and not depending on whether the `metricsConfig.enabled` is set. The major advantage is that liveness and readiness probes are always enabled.

ChiefAlexander replied:
I missed that the probes relayed on metrics. I agree this should always be on so that the pods have their probes to determine health.

ChiefAlexander commented on charts/argo-workflows/templates/controller/workflow-controller-deployment.yaml:
```diff
@@ -0,0 +1,114 @@
+          livenessProbe:
+            httpGet:
+              port: metrics
+              path: {{- .Values.controller.metricsConfig.path }}
```
```suggestion
              path: {{ .Values.controller.metricsConfig.path }}
```

ChiefAlexander commented on charts/argo-workflows/templates/controller/workflow-controller-deployment.yaml:
```diff
@@ -0,0 +1,114 @@
+          readinessProbe:
+            httpGet:
+              port: metrics
+              path: {{- .Values.controller.metricsConfig.path }}
```
```suggestion
              path: {{ .Values.controller.metricsConfig.path }}
```

psmit commented on charts/argo-workflows/Chart.yaml:
```diff
@@ -0,0 +1,14 @@
+apiVersion: v2
+name: argo-workflows
```
Just out of curiosity; you are not defining an appVersion and set the tag right away in the values.yaml. Does this have a specific reason?

chgl replied:
The main motivation (or why I am personally always doing it) is to support automated image tag updates in the future, see https://github.com/argoproj/argo-helm/issues/536. If all container images are set consistently from within the values.yaml, then updates are possible out-of-the-box (at least for renovatebot should we ever consider using it).

It might make sense to set `appVersion: "v3"`?

psmit replied:
I see that tools like renovatebot might have trouble with the appVersion; but I believe it would be "the other way around" to now for this chart to not adopt the industry-standard practices, instead of finding a way for the other tools to support the appVersion attribute.

I would propose to use the helm-standard "appVersion" to contain the argo version and let image.tag default to .Chart.AppVersion

chgl replied:
I'm not completely convinced yet, but since my use case is at this point entirely hypothetical I have reverted back to using appVersion.

psmit commented on charts/argo-workflows/Chart.yaml:
```diff
@@ -0,0 +1,14 @@
+description: A Helm chart for Argo Workflows
+type: application
+version: 0.1.0
+icon: https://raw.githubusercontent.com/argoproj/argo/master/docs/assets/argo.png
```
This link should be updated to https://raw.githubusercontent.com/argoproj/argo-workflows/master/docs/assets/argo.png

psmit commented on charts/argo-workflows/templates/NOTES.txt:
```diff
@@ -0,0 +1,7 @@
+
+2. Submit the hello-world workflow by running:
+
+argo submit https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml --watch
```
This link should be updated to https://raw.githubusercontent.com/argoproj/argo-workflows/master/examples/hello-world.yaml

psmit commented on charts/argo-workflows/values.yaml:
```diff
@@ -0,0 +1,325 @@
+  # ref: https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/
+  # PriorityClass: system-cluster-critical
+  priorityClassName: ""
+  # https://argoproj.github.io/argo/links/
```
This link needs to be updated to https://argoproj.github.io/argo-workflows/links/

psmit commented on charts/argo-workflows/values.yaml:
```diff
@@ -0,0 +1,325 @@
+  enabled: true
+  # only updates base url of resources on client side,
+  # it's expected that a proxy server rewrites the request URL and gets rid of this prefix
+  # https://github.com/argoproj/argo/issues/716#issuecomment-433213190
```
Should be updated to https://github.com/argoproj/argo-workflows/issues/716#issuecomment-433213190

chgl commented on charts/argo-workflows/templates/controller/workflow-controller-deployment.yaml:
```diff
@@ -0,0 +1,106 @@
+metadata:
+  name: {{ .Release.Name }}-{{ .Values.controller.name}}
+  labels:
+    app: {{ .Release.Name }}-{{ .Values.controller.name}}
```
the deployments are currently using the "old-style" labels while newer deployments/charts generally use the recommended labels such as `app.kubernetes.io/name` (see https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/)

Should we update these as well?

oliverbaehler replied:
I was wondering if we should add a library chart that has the same labels/ selectors etc. for all charts. We recently implemented the new-style labels in the argo-cd chart. I would suggest for now we leave it old style but i will create an issue addressing this for all charts.

chgl replied:
Sounds good!

ChiefAlexander replied:
I'd personally like to see these updated to the recommended style. However doing so is a breaking change as selector labels are immutable.

chgl replied:
You're right! I'm OK with either approach. I've looked at https://github.com/argoproj/argo-helm/blob/master/charts/argo-cd/templates/_helpers.tpl and could definitely copy-and-paste them over to this chart for now until we have a library chart available. Since this is the first version of this chart and a 0.* release, most people will need to freshly install anyway.

chgl commented:
I can't access the circleci logs unfortunately, but running the lint.sh script locally yields `Error: Looks like "https://kubernetes-charts.storage.googleapis.com" is not a valid chart repository or cannot be reached: Failed to fetch https://kubernetes-charts.storage.googleapis.com/index.yaml : 403 Forbidden` - running `helm init --stable-repo-url=https://charts.helm.sh/stable --client-only` seems to fix it though. But since other PRs are fine I think I'm doing something else wrong.

oliverbaehler commented:
@chgl You should be able to view the circle-ci logs. Click on details and then login via Github. It requires a version bump of the old `argo` chart:

```
 ✖︎ argo => (version: "0.16.9", path: "charts/argo") > Chart version not ok. Needs a version bump!
``` 
Well that's what chart-linter is causing to fail.

chgl commented:
Ah, thank you! I just linted argo-workflows instead of the entire charts/ dir locally. 

I'm a bit hesitant to grant circleci access, so I'm definitely looking forward to your PR https://github.com/argoproj/argo-helm/pull/685 being merged!

oliverbaehler commented:
@chgl Understandable. Doesn't look like anyone wants to drive it forward but let's hope for the best ^^

ionutz89 commented:
Can we merge this in master? I am getting issue with the current helm chart for workflow. It seems the dependencies of minio helm chart makes deploy unsuccessful.  I tested this PR and it is working for me as a new fresh install via ArgoCD.

oliverbaehler commented:
@ionutz89 I will try to review and merge it in the next days

oliverbaehler commented on charts/argo-workflows/templates/server/server-deployment.yaml:
```diff
@@ -0,0 +1,104 @@
+{{ toYaml .Values.server.podAnnotations | indent 8}}{{- end }}
+    spec:
+      serviceAccountName: {{ .Values.server.serviceAccount | quote }}
+      {{- if .Values.server.podSecurityContext }}
+      securityContext:
+        {{- toYaml .Values.server.podSecurityContext | nindent 8 }}
+      {{- end }}
```
```suggestion
      {{- with .Values.server.podSecurityContext }}
      securityContext:
        {{- toYaml . | nindent 8 }}
      {{- end }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-deployment.yaml:
```diff
@@ -0,0 +1,104 @@
+          args:
+          - server
+          - --configmap={{ .Release.Name }}-{{ .Values.controller.name }}-configmap
+          {{- if .Values.server.extraArgs }}
+          {{- toYaml .Values.server.extraArgs | nindent 10 }}
+          {{- end }}
```
```suggestion
          {{- with .Values.server.extraArgs }}
            {{- toYaml . | nindent 10 }}
          {{- end }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-deployment.yaml:
```diff
@@ -0,0 +1,104 @@
+      affinity:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- if .Values.server.priorityClassName }}
+      priorityClassName: {{ .Values.server.priorityClassName }}
```
```suggestion
      {{- with .Values.server.priorityClassName }}
      priorityClassName: {{ . }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-deployment.yaml:
```diff
@@ -0,0 +1,104 @@
+      labels:
+        {{- include "argo-workflows.labels" (dict "context" . "component" .Values.server.name "name" .Values.server.name) | nindent 8 }}
+        app.kubernetes.io/version: {{ default .Chart.AppVersion .Values.server.image.tag | quote }}
+        {{- if .Values.server.podLabels }}
+        {{- toYaml .Values.server.podLabels | nindent 8 }}
+        {{- end }}
```
```suggestion
        {{- with .Values.server.podLabels }}
          {{- toYaml . | nindent 8 }}
        {{- end }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-sa.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: ServiceAccount
+metadata:
+  name: {{ template "argo-workflows.serverServiceAccountName" . }}
+  annotations:
+{{ toYaml .Values.server.serviceAccountAnnotations | indent 4 }}
```
```suggestion
  {{- with .Values.server.serviceAccountAnnotations  }}
  annotations:
    {{- toYaml . | indent 4 }}
  {{- end }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-service.yaml:
```diff
@@ -0,0 +1,29 @@
+  labels:
+    {{- include "argo-workflows.labels" (dict "context" . "component" .Values.server.name "name" .Values.server.name) | nindent 4 }}
+    app.kubernetes.io/version: {{ default .Chart.AppVersion .Values.server.image.tag | quote }}
+  {{- if .Values.server.serviceAnnotations }}
+  annotations:
+{{ toYaml .Values.server.serviceAnnotations | indent 4}}{{- end }}
```
```suggestion
  {{- with .Values.server.serviceAnnotations }}
  annotations:
    {{- toYaml . | nindent 4 }}
  {{- end }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-service.yaml:
```diff
@@ -0,0 +1,29 @@
+  loadBalancerIP: {{ .Values.server.loadBalancerIP | quote }}
+  {{- end }}
+  {{- if and (eq .Values.server.serviceType "LoadBalancer") .Values.server.loadBalancerSourceRanges }}
+  loadBalancerSourceRanges:
+{{ toYaml .Values.server.loadBalancerSourceRanges | indent 4 }}{{- end }}
```
```suggestion
  loadBalancerSourceRanges:
    {{- toYaml .Values.server.loadBalancerSourceRanges | nindent 4 }}
  {{- end }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-service.yaml:
```diff
@@ -0,0 +1,29 @@
+spec:
+  ports:
+  - port: {{ .Values.server.servicePort }}
+    {{- if .Values.server.servicePortName }}
+    name: {{ .Values.server.servicePortName }}
```
```suggestion
    {{- with .Values.server.servicePortName }}
    name: {{ . }}
```

oliverbaehler commented on charts/argo/Chart.yaml:
```diff
@@ -5,13 +5,9 @@ name: argo
 version: 0.16.10
```
We also bumped to a new major when deprecating argo-ci, so i would do the same here

```suggestion
version: 1.0.0
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,39 @@
+              serviceName: {{ $serviceName }}
+              servicePort: {{ $servicePort }}
+    {{- end -}}
+  {{- if .Values.server.ingress.tls }}
+  tls:
+{{ toYaml .Values.server.ingress.tls | indent 4 }}
+  {{- end -}}
+{{- end -}}
```
```suggestion
    {{- with .Values.server.ingress.tls }}
  tls:
      {{- toYaml . | nindent 4 }}
    {{- end -}}
  {{- end -}}
```

oliverbaehler commented on charts/argo-workflows/templates/controller/workflow-controller-servicemonitor.yaml:
```diff
@@ -0,0 +1,29 @@
+      path: {{ .Values.controller.metricsConfig.path }}
+      interval: 30s
+  {{- end }}
+  {{- if .Values.controller.telemetryConfig.enabled }}
+    - port: telemetry
+      path: {{ .Values.controller.telemetryConfig.path }}
```
```suggestion
  {{- with .Values.controller.telemetryConfig.enabled }}
    - port: telemetry
      path: {{ . }}
```

oliverbaehler commented on charts/argo-workflows/templates/controller/workflow-controller-servicemonitor.yaml:
```diff
@@ -0,0 +1,29 @@
+    {{- end }}
+spec:
+  endpoints:
+  {{- if .Values.controller.metricsConfig.enabled }}
+    - port: metrics
+      path: {{ .Values.controller.metricsConfig.path }}
```
```suggestion
  {{- with .Values.controller.metricsConfig.enabled }}
    - port: metrics
      path: {{ . }}
```

oliverbaehler commented on charts/argo-workflows/templates/controller/workflow-controller-sa.yaml:
```diff
@@ -0,0 +1,6 @@
+kind: ServiceAccount
+metadata:
+  name: {{ template "argo-workflows.controllerServiceAccountName" . }}
+  annotations:
+{{ toYaml .Values.controller.serviceAccountAnnotations | indent 4 }}
```
```suggestion
  {{ with .Values.controller.serviceAccountAnnotations }}
  annotations:
    {{- toYaml .| nindent 4 }}
  {{- end }}
```

oliverbaehler commented on charts/argo-workflows/templates/controller/workflow-controller-deployment.yaml:
```diff
@@ -0,0 +1,103 @@
+{{ toYaml .Values.controller.podAnnotations | indent 8}}{{- end }}
+    spec:
+      serviceAccountName: {{ template "argo-workflows.controllerServiceAccountName" . }}
+      {{- if .Values.controller.podSecurityContext }}
+      securityContext:
+        {{- toYaml .Values.controller.podSecurityContext | nindent 8 }}
+      {{- end }}
```
```suggestion
      {{- with .Values.controller.podSecurityContext }}
      securityContext:
        {{- toYaml . | nindent 8 }}
      {{- end }}
```

oliverbaehler reviewed (changes requested):
@chgl I hope you don't hate me too much for this. I have corrected all the spots where if was used instead of `with` to keep a better consistency. I corrected some indendting and replaced indent with nindent. I think there's one proposal in the rbac which changes coniditions but the reast is just cleaning up. I thought it would be a good opportunity and further contributions now face a cleaner code. What do you think?

oliverbaehler commented:
@chgl Also, don't you think ye just should keep the minio dependency for now? As mentioned in the related pr to the issue there's a demand for a minio tenant chart. But they are still missing stuff like default Bucketsetc. on minio tenants. And I would wait minio releases a tenant chart themselves. As far as I am concerned I would keep the minio dependency.

@ionutz89 What problem are you seeing with the minio dependency? Because currently you won't  have a minio dependency at all, which is also already supported via values

ionutz89 commented:
> @ionutz89 What problem are you seeing with the minio dependency? Because currently you won't have a minio dependency at all, which is also already supported via values

@oliverbaehler I am getting this error:
`rpc error: code = Unknown desc = helm dependency build failed exit status 1: Error: no repository definition for https://helm.min.io/. Please add the missing repos via 'helm repo add'` via ArgoCD install with all defaults settings. 

In value.yaml file we have this currently:
```
minio:
  install: false
  ```
  
 It's set to false. I tried to overwrite this value but no luck.

oliverbaehler commented:
@ionutz89 I am aware of that, since I am working on fixing this lifecycle. However this does not instantly deprecate the old dependency and we are not gonna deploy an operator as dependency of argo. Before the following issues aren't handled by minio we are not gonna change to the operator based minio instances.

> via ArgoCD install with all defaults settings

Not exactly sure what that means, but for me it works fine with deploying the default values (there's no reason to contact the minio repository with the default values):

```
 helm install my-argo argo/argo --version 0.16.10 -n argo-work
W0427 21:40:19.627094  232593 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:40:19.632717  232593 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:40:19.638812  232593 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:40:19.644851  232593 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:40:19.650380  232593 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
NAME: my-argo
LAST DEPLOYED: Tue Apr 27 21:40:19 2021
NAMESPACE: argo-work
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
1. Get Argo Server external IP/domain by running:

kubectl --namespace argo-work get services -o wide | grep my-argo-server

2. Submit the hello-world workflow by running:

argo submit https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml --watc
```

And with Minio enabled:

``` 
W0427 21:42:45.244461  242888 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:42:45.250265  242888 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:42:45.256166  242888 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:42:45.262123  242888 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 21:42:45.268457  242888 warnings.go:67] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
NAME: my-argo
LAST DEPLOYED: Tue Apr 27 21:42:45 2021
NAMESPACE: argo-work
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
1. Get Argo Server external IP/domain by running:

kubectl --namespace argo-work get services -o wide | grep my-argo-server

2. Submit the hello-world workflow by running:

argo submit https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml --watch



3. Access Minio UI and create bucket 'argo-artifacts'. Minio UI is available on port 9000 and available via external URL. URL might be retrieved using following
command:

kubectl --namespace argo-work get services -o wide | grep my-argo-minio
```
So your problem seems to be somewhere else.

ionutz89 commented:
@oliverbaehler thank you for checking out. Appreciated your very quick reply. 🍺 

The output from the previous command was from the `ArgoCD UI` interface.
If it is required, I can provide screenshot.

This is config.yaml file tested:

```
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: argocd-workflow
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  destination:
    namespace: argocd
    server: https://kubernetes.default.svc
  project: default
  source:
    path: charts/argo
    repoURL: https://github.com/argoproj/argo-helm.git
  syncPolicy:
    automated:
      selfHeal: true
```
I am just creating a new app via Argo CD with, not via CLI. So ArgoCD handles all helm installations. 
Via CLI is working fine even with your command. So I am not sure what I am doing wrong. 🤔 

````
 helm install my-argo argo/argo --version 0.16.10 -n argo-work
W0427 20:54:08.811837 2398436 warnings.go:70] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 20:54:08.877059 2398436 warnings.go:70] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 20:54:08.958523 2398436 warnings.go:70] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 20:54:09.032538 2398436 warnings.go:70] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
W0427 20:54:09.076939 2398436 warnings.go:70] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
manifest_sorter.go:192: info: skipping unknown hook: "crd-install"
NAME: my-argo
LAST DEPLOYED: Tue Apr 27 20:54:16 2021
NAMESPACE: argo-work
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
1. Get Argo Server external IP/domain by running:

kubectl --namespace argo-work get services -o wide | grep my-argo-server

2. Submit the hello-world workflow by running:

argo submit https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml --watch
❯ kubectl  get po -n argo-work
NAME                                           READY   STATUS              RESTARTS   AGE
my-argo-server-7c695959c6-frdff                0/1     ContainerCreating   0          16s
my-argo-workflow-controller-6cb5797dcd-q6d4j   0/1     ContainerCreating   0          16s

oliverbaehler commented:
@ionutz89 Oh I see. You need to add minio as helm repository when deploying that way. See When you build it from the git repository, the dependency is missing and therefor it tries building it (And that's why you are getting that error message). But with the helm packages we release (and are installed via cli) the reference to the helm repository is hardcoded and therefor it's always able to fetch the chart.

What you need to do is adding the minio repository (based on [argo-cd repositories values](https://github.com/argoproj/argo-helm/blob/master/charts/argo-cd/values.yaml#L571)) would look something like this:

```
...
repositories: |
  - type: helm
    url: https://helm.min.io/
    name: minio
....
```

Try that and it should work. Another way would be instead of referencing the Git Repository the Helm Repository in your deployed application.

bmayfi3ld commented on charts/argo-workflows/templates/_helpers.tpl:
```diff
@@ -0,0 +1,96 @@
+{{/*
+Return the appropriate apiVersion for ingress
+*/}}
+{{- define "argo-cd.ingress.apiVersion" -}}
```
Should this be 

```suggestion
{{- define "argo-workflows.ingress.apiVersion" -}}
```

I wasn't able to get it to work otherwise with below error.

`UPGRADE FAILED: template: argo-workflows/templates/server/server-ingress.yaml:5:15: executing \"argo-workflows/templates/server/server-ingress.yaml\" at <include \"argo-workflows.ingress.apiVersion\" .>: error calling include: template: no template \"argo-workflows.ingress.apiVersion\" associated with template \"gotpl\"\n",`

server-ingress.yaml references it as

chgl replied:
That's embarassing, you are correct of course!

bmayfi3ld commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,39 @@
+metadata:
+  name: {{ template "argo-workflows.server.fullname" . }}
+  labels:
+    {{- include "argo-workflows.labels" $ | nindent 4 }}
```
```suggestion
   {{- include "argo-workflows.labels" (dict "context" . "component" .Values.server.name "name" .Values.server.name) | nindent 4 }}
```

Making this same as other references to this template. Was getting an error otherwise.
error calling include: template: argo-workflows/templates/_helpers.tpl:57:39: executing \"argo-`workflows.selectorLabels\" at <.context.Release.Name>: nil pointer evaluating interface {}.Release"`

bmayfi3ld reviewed (commented):
Didn't mean to make this a review, but similar error as in the template file.

bmayfi3ld commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,39 @@
+{{- $serviceName := include "argo-workflows.server.fullname" . -}}
+{{- $servicePort := .Values.server.servicePort -}}
+apiVersion: {{ include "argo-workflows.ingress.apiVersion" . }}
+metadata:
```
```suggestion
kind: Ingress
metadata:
```

bmayfi3ld commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,39 @@
+    {{- end }}
+spec:
+  rules:
+    {{- range .Values.server.ingress.hosts }}
+    - host: {{ . }}
+      http:
+        paths:
+          {{- if $.Values.server.ingress.paths }}
+          {{- range $.Values.server.ingress.paths }}
+          - backend:
+              serviceName: {{ .serviceName }}
+              servicePort: {{ .servicePort }}
+          {{- end }}
+          {{- end }}
+          - backend:
+              serviceName: {{ $serviceName }}
+              servicePort: {{ $servicePort }}
```
```suggestion
    {{- $kubeVersion := .Capabilities.KubeVersion.GitVersion}}
    {{- range .Values.server.ingress.hosts }}
    - host: {{ . }}
      http:
        paths:
          {{- if $.Values.server.ingress.paths }}
          {{- range $.Values.server.ingress.paths }}
          {{- include "argo-workflows.ingress.service" (dict "kubeVersion" $kubeVersion "serviceName" .serviceName "servicePort" .servicePort) | nindent 10 }}
          {{- end }}
          {{- end }}
          {{- include "argo-workflows.ingress.service" (dict "kubeVersion" $kubeVersion "serviceName" $serviceName "servicePort" $servicePort) | nindent 10 }}
```

Also see new helper function. The new ingress api changed the fields layout. Since the api field in here was attempting to be backward compatible. I suggested a function to correctly build the backend field.

bmayfi3ld commented on charts/argo-workflows/templates/_helpers.tpl:
```diff
@@ -0,0 +1,96 @@
+{{- else -}}
+{{- print "networking.k8s.io/v1" -}}
+{{- end -}}
+{{- end -}}
```
```suggestion
{{- end -}}

{{/*
Return the appropriate service layout for ingress
*/}}
{{- define "argo-workflows.ingress.service" -}}
{{- if semverCompare "<1.19-0" .kubeVersion -}}
- backend:
    serviceName: {{ .serviceName }}
    servicePort: {{ .servicePort }}
{{- else -}}
- backend:
    service:
      name: {{ .serviceName }}
      port: 
        number: {{ .servicePort }}
  pathType: ImplementationSpecific
{{- end -}}
{{- end -}}
```

To compliment fixing errors in ingress.

bmayfi3ld commented:
Added several changes for the server's ingress resource. It didn't seem to be functional. But I had been using it in the argo chart.

chgl commented:
> Added several changes for the server's ingress resource. It didn't seem to be functional. But I had been using it in the argo chart.

@bmayfi3ld Thank you! I've now re-used the ingress definition from the argo-cd chart and also added a values file for chart-testing.

bmayfi3ld commented on charts/argo-workflows/templates/server/server-deployment.yaml:
```diff
@@ -0,0 +1,105 @@
+            {{- toYaml .Values.server.securityContext | nindent 12 }}
+          args:
+          - server
+          - --configmap={{ .Release.Name }}-{{ .Values.controller.name }}-configmap
```
```suggestion
          - --configmap={{ template "argo-workflows.controller.fullname" . }}-configmap
```

I believe this is the name of the configmap

oliverbaehler commented:
@chgl @bmayfi3ld We encounter currently a problem with said ingress condition in argo-cd #703. Currently my best idea would be the add a global value which allows to overwrite the kubeVersion. But i might have to sleep over it.

tyrken commented on charts/argo-workflows/templates/controller/workflow-controller-servicemonitor.yaml:
```diff
@@ -0,0 +1,29 @@
+  endpoints:
+  {{- with .Values.controller.metricsConfig.enabled }}
+    - port: metrics
+      path: {{ . }}
```
The path value should be from `.Values.controller.metricsConfig.path` - not the `true` that comes from the `with` expression.  Not sure if `with` really helps here or if it's better going back to the old fashioned `if`.

chgl replied:
thanks! Agreed, here an `if` is definitely clearer.

tyrken commented on charts/argo-workflows/templates/controller/workflow-controller-servicemonitor.yaml:
```diff
@@ -0,0 +1,29 @@
+  {{- end }}
+  {{- with .Values.controller.telemetryConfig.enabled }}
+    - port: telemetry
+      path: {{ . }}
```
Same `path` issue as above, but needs to be `.Values.controller.telemetryConfig.path`

tyrken commented:
Just found a couple of minor issues in the controller's ServiceMonitor, commented inline.  Otherwise looks good & works when I disabled them.  Would be nice to bump up the appVersion to `3.0.2` given it's been out a couple of weeks now.

Thanks to all for your efforts!

chgl commented:
Thank you @tyrken! I've corrected it and also moved the previously known as `telemetryServicePort` inside the `telemetryConfig` as `telemetryConfig.servicePort` - same for `metricsConfig`.

I've additionally aligned the configuration of serviceAccounts with the argo-cd chart, ie: what used to be `server.createServiceAccount` is now `server.serviceAccount.create`, etc.  I've also found and corrected a few places in the RoleBindings where I hadn't used the `server/controllerServiceAccountName` templates.

abhinavkhanna-sf commented:
Thank you @chgl for the PR. I am actually tracking this PR everyday.
it still shows changes request by @oliverbaehler . Is it ready for review again?

abhinavkhanna-sf commented:
@oliverbaehler can you please review this PR?

oliverbaehler commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,88 @@
+            pathType: Prefix
+            {{- end }}
+            backend:
+              {{- if eq (include "argo-workflows.ingress.apiVersion" $) "networking.k8s.io/v1" }}
```
```suggestion
              {{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1" }}
```

chgl replied:
Thanks! I added `/Ingress"` to make sure it checks for the availability of the Ingress version specifically - there was a comment here somewhere indicating the need for it.

oliverbaehler commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,88 @@
+          {{- end }}
+          {{- range $p := $paths }}
+          - path: {{ $p }}
+            {{- if eq (include "argo-workflows.ingress.apiVersion" $) "networking.k8s.io/v1" }}
```
```suggestion
            {{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1" }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,88 @@
+            pathType: Prefix
+            {{- end }}
+            backend:
+              {{- if eq (include "argo-workflows.ingress.apiVersion" $) "networking.k8s.io/v1" }}
```
```suggestion
              {{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1" }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,88 @@
+          {{- end }}
+          {{- range $p := $paths }}
+          - path: {{ $p }}
+            {{- if eq (include "argo-workflows.ingress.apiVersion" $) "networking.k8s.io/v1" }}
```
```suggestion
            {{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1" }}
```

oliverbaehler commented on charts/argo-workflows/templates/server/server-ingress.yaml:
```diff
@@ -0,0 +1,88 @@
+    {{- toYaml .Values.server.ingress.labels | nindent 4 }}
+    {{- end }}
+spec:
+  {{- if eq (include "argo-workflows.ingress.apiVersion" $) "networking.k8s.io/v1" }}
```
```suggestion
  {{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1" }}
```

oliverbaehler commented on charts/argo-workflows/templates/_helpers.tpl:
```diff
@@ -0,0 +1,114 @@
+{{- end -}}
+{{- end -}}
+
+{{/*
+Return the appropriate service layout for ingress
+*/}}
+{{- define "argo-workflows.ingress.service" -}}
+{{- if semverCompare "<1.19-0" .kubeVersion -}}
+- backend:
+    serviceName: {{ .serviceName }}
+    servicePort: {{ .servicePort }}
+{{- else -}}
+- backend:
+    service:
+      name: {{ .serviceName }}
+      port:
+        number: {{ .servicePort }}
+  pathType: ImplementationSpecific
+{{- end -}}
+{{- end -}}
```
Remove this

oliverbaehler reviewed (changes requested):
@chgl Final Changes, after that I will be satisfied and further fixes changes should be done with additional prs, because we are kinda losing the overview here. Anyway, thanks for the work!

abhinavkhanna-sf commented:
@oliverbaehler Can you please review this ?

oliverbaehler reviewed (approved):
lgtm

oliverbaehler commented:
@chgl Thanks for all your efforts! :heart: 
The Chart is now also available on Artifacthub: https://artifacthub.io/packages/helm/argo/argo-workflows

Please test & open new prs if any errors occur.

chgl commented:
Thanks again @oliverbaehler for your time (and patience) reviewing this PR!
